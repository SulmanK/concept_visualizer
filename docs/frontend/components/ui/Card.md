# Card Component

The `Card` component is a flexible container that displays content with a consistent style, providing a clean and organized visual representation of information.

## Overview

The Card component extends Material-UI's Card component with additional styling and functionality, making it easier to create consistently styled content containers throughout the application. It provides options for elevation, hover effects, and click interactions.

## Usage

```tsx
import { Card } from 'components/ui/Card';

// Basic usage
<Card>
  <h3>Card Title</h3>
  <p>Card content goes here</p>
</Card>

// With hover and click functionality
<Card 
  hoverEffect 
  onClick={() => console.log('Card clicked')}
>
  <h3>Interactive Card</h3>
  <p>Click me!</p>
</Card>
```

## Props

| Prop | Type | Default | Description |
|------|------|---------|-------------|
| `children` | `React.ReactNode` | - | Content to render inside the card |
| `className` | `string` | `''` | Additional CSS class to apply to the card |
| `elevation` | `number` | `1` | Shadow depth, higher values mean more raised appearance |
| `hoverEffect` | `boolean` | `false` | Whether to apply elevation increase on hover |
| `onClick` | `() => void` | - | Function to call when the card is clicked |
| `sx` | `SxProps<Theme>` | `{}` | MUI system props for additional styling |
| `variant` | `'outlined' \| 'elevation'` | `'elevation'` | Card variant style |

## Implementation Details

```tsx
import React from 'react';
import { 
  Card as MuiCard, 
  CardProps as MuiCardProps, 
  styled 
} from '@mui/material';
import { SxProps, Theme } from '@mui/material/styles';

interface CardProps extends Omit<MuiCardProps, 'onClick'> {
  children: React.ReactNode;
  className?: string;
  elevation?: number;
  hoverEffect?: boolean;
  onClick?: () => void;
  sx?: SxProps<Theme>;
  variant?: 'outlined' | 'elevation';
}

const StyledCard = styled(MuiCard, {
  shouldForwardProp: (prop) => prop !== 'hoverEffect',
})<{ hoverEffect?: boolean }>(({ theme, hoverEffect }) => ({
  transition: theme.transitions.create(['box-shadow', 'transform'], {
    duration: theme.transitions.duration.shorter,
  }),
  ...(hoverEffect && {
    '&:hover': {
      transform: 'translateY(-4px)',
      boxShadow: theme.shadows[4],
    },
  }),
}));

export function Card({
  children,
  className = '',
  elevation = 1,
  hoverEffect = false,
  onClick,
  sx = {},
  variant = 'elevation',
  ...rest
}: CardProps) {
  return (
    <StyledCard
      className={className}
      elevation={elevation}
      hoverEffect={hoverEffect}
      onClick={onClick}
      sx={{
        cursor: onClick ? 'pointer' : 'default',
        ...sx,
      }}
      variant={variant}
      {...rest}
    >
      {children}
    </StyledCard>
  );
}
```

## Features

- **Consistent Styling**: Follows the application's design system
- **Interactive Options**: Supports hover effects and click handlers
- **Flexible Content**: Can contain any type of content
- **Customizable Appearance**: Adjustable elevation and variant
- **Styling Extensions**: Supports MUI's `sx` prop for additional styling
- **Accessibility**: Maintains ARIA roles and states when interactive

## Usage Patterns

### Basic Information Card

```tsx
<Card>
  <CardContent>
    <Typography variant="h5" component="div">
      Card Title
    </Typography>
    <Typography variant="body2" color="text.secondary">
      Card description text with details about the content.
    </Typography>
  </CardContent>
</Card>
```

### Interactive Card with Image

```tsx
<Card 
  hoverEffect 
  onClick={() => navigate(`/concepts/${concept.id}`)}
  sx={{ maxWidth: 345 }}
>
  <CardMedia
    component="img"
    height="140"
    image={concept.imageUrl}
    alt={concept.name}
  />
  <CardContent>
    <Typography variant="h6">{concept.name}</Typography>
    <Typography variant="body2" color="text.secondary">
      {concept.description}
    </Typography>
  </CardContent>
</Card>
```

### Outlined Card

```tsx
<Card variant="outlined" sx={{ borderColor: 'primary.light' }}>
  <CardContent>
    <Typography variant="h6">Information Panel</Typography>
    <Typography variant="body2">
      Important information displayed in an outlined card.
    </Typography>
  </CardContent>
</Card>
```

### Card with Action Buttons

```tsx
<Card>
  <CardContent>
    <Typography variant="h6">Feature Card</Typography>
    <Typography variant="body2">
      Feature description and details.
    </Typography>
  </CardContent>
  <CardActions>
    <Button size="small">Learn More</Button>
    <Button size="small" color="primary">
      Try Now
    </Button>
  </CardActions>
</Card>
```

## Best Practices

1. Use cards to group related information and actions
2. Apply the `hoverEffect` only when the card is interactive (has an `onClick` handler)
3. Maintain consistent padding and spacing within cards
4. Use appropriate elevation levels based on the card's importance (higher elevation for more important content)
5. Include clear call-to-action elements in interactive cards 