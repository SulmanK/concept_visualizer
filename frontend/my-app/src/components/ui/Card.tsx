import React, { HTMLAttributes } from "react";
import { usePrefersReducedMotion } from "../../hooks";

export interface CardProps extends HTMLAttributes<HTMLDivElement> {
  /**
   * Card variant
   */
  variant?: "default" | "gradient" | "elevated";

  /**
   * Header content
   */
  header?: React.ReactNode;

  /**
   * Footer content
   */
  footer?: React.ReactNode;

  /**
   * Add extra padding to the card
   */
  padded?: boolean;

  /**
   * Card is in loading state
   */
  isLoading?: boolean;

  /**
   * Whether the card should have hover effects
   * @default false
   */
  interactive?: boolean;

  /**
   * Type of hover animation
   * @default 'lift'
   */
  hoverEffect?: "lift" | "glow" | "border" | "scale" | "none";
}

/**
 * Card component for containing content
 */
export const Card: React.FC<CardProps> = ({
  variant = "default",
  header,
  footer,
  padded = true,
  isLoading = false,
  interactive = true,
  hoverEffect = "lift",
  className = "",
  children,
  ...props
}) => {
  const prefersReducedMotion = usePrefersReducedMotion();

  // Base card styles for each variant
  const baseClasses = {
    default:
      "bg-white/90 backdrop-blur-sm rounded-lg shadow-modern border border-indigo-100 overflow-hidden",
    gradient:
      "bg-gradient-to-br from-indigo-50/90 to-white/90 backdrop-blur-sm rounded-lg shadow-modern border border-indigo-100 overflow-hidden",
    elevated: "bg-white border-none rounded-lg shadow-lg overflow-hidden",
  };

  // Transition classes (skip if reduced motion is preferred)
  const transitionClass = !prefersReducedMotion
    ? "transition-all duration-300"
    : "";

  // Classes for different hover effects when card is interactive
  const getHoverClasses = () => {
    if (!interactive || prefersReducedMotion) return "";

    switch (hoverEffect) {
      case "lift":
        return "hover:translate-y-[-4px] hover:shadow-lg";
      case "glow":
        return "hover:shadow-[0_0_15px_rgba(79,70,229,0.3)]";
      case "border":
        return "hover:border-indigo-300";
      case "scale":
        return "hover:scale-[1.02]";
      case "none":
        return "";
      default:
        return "";
    }
  };

  // Combine all card classes
  const cardClass = `${
    baseClasses[variant]
  } ${transitionClass} ${getHoverClasses()} ${className}`.trim();

  // Content, header and footer classes
  const contentClass = padded ? "p-4 sm:p-6" : "";
  const headerClass =
    "px-4 py-3 sm:px-6 border-b border-indigo-200 bg-indigo-50/50";
  const footerClass =
    "px-4 py-3 sm:px-6 border-t border-indigo-200 bg-indigo-50/50";

  return (
    <div className={cardClass} {...props}>
      {header && <div className={headerClass}>{header}</div>}

      <div className={contentClass}>
        {isLoading ? (
          <div className="flex justify-center items-center py-12">
            <div className="animate-spin rounded-full h-12 w-12 border-t-2 border-b-2 border-primary"></div>
          </div>
        ) : (
          children
        )}
      </div>

      {footer && <div className={footerClass}>{footer}</div>}
    </div>
  );
};

// Add default export to fix import issue
export default Card;
