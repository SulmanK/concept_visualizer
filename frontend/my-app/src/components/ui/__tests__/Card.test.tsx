import React from 'react';
import { render, screen } from '@testing-library/react';
import { Card } from '../Card';

describe('Card Component', () => {
  // Basic rendering tests
  test('renders card with children', () => {
    render(
      <Card>
        <p>Card content</p>
      </Card>
    );
    
    const content = screen.getByText('Card content');
    expect(content).toBeInTheDocument();
  });
  
  // Variant tests
  test.each([
    ['default', 'card'],
    ['gradient', 'card-gradient'],
    ['elevated', 'card bg-white border-none shadow-lg'],
  ])('renders %s variant correctly', (variant, expectedClass) => {
    render(
      <Card variant={variant as any}>
        <p>Card content</p>
      </Card>
    );
    
    const contentElement = screen.getByText('Card content');
    const card = contentElement.parentElement?.parentElement;
    expect(card?.className).toContain(expectedClass.split(' ')[0]); // Just check the first class
  });
  
  // Padding tests
  test('renders card with padding when padded is true', () => {
    render(
      <Card padded>
        <p>Card content</p>
      </Card>
    );
    
    const contentDiv = screen.getByText('Card content').parentElement;
    expect(contentDiv?.className).toContain('p-4');
  });
  
  test('renders card without padding when padded is false', () => {
    render(
      <Card padded={false}>
        <p>Card content</p>
      </Card>
    );
    
    const contentDiv = screen.getByText('Card content').parentElement;
    expect(contentDiv?.className).toBe('');
  });
  
  // Loading state tests
  test('renders loading spinner when isLoading is true', () => {
    render(
      <Card isLoading>
        <p>Card content</p>
      </Card>
    );
    
    const spinner = document.querySelector('.animate-spin');
    expect(spinner).not.toBeNull();
  });
  
  // Custom className test
  test('applies custom className when provided', () => {
    render(
      <Card className="custom-class">
        <p>Card content</p>
      </Card>
    );
    
    const contentElement = screen.getByText('Card content');
    const card = contentElement.parentElement?.parentElement;
    expect(card?.className).toContain('custom-class');
  });
  
  // Header and footer tests
  test('renders card with header when provided', () => {
    const header = <div data-testid="card-header">Custom Header</div>;
    
    render(
      <Card header={header}>
        <p>Card content</p>
      </Card>
    );
    
    const cardHeader = screen.getByTestId('card-header');
    expect(cardHeader).toBeInTheDocument();
    expect(cardHeader.textContent).toBe('Custom Header');
  });
  
  test('renders card with footer when provided', () => {
    const footer = <div data-testid="card-footer">Custom Footer</div>;
    
    render(
      <Card footer={footer}>
        <p>Card content</p>
      </Card>
    );
    
    const cardFooter = screen.getByTestId('card-footer');
    expect(cardFooter).toBeInTheDocument();
    expect(cardFooter.textContent).toBe('Custom Footer');
  });
  
  // Snapshot tests
  describe('Snapshots', () => {
    test('default card snapshot', () => {
      const { container } = render(
        <Card>
          <p>Default card content</p>
        </Card>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
    
    test('gradient variant card snapshot', () => {
      const { container } = render(
        <Card variant="gradient">
          <p>Gradient card content</p>
        </Card>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
    
    test('elevated variant card snapshot', () => {
      const { container } = render(
        <Card variant="elevated">
          <p>Elevated card content</p>
        </Card>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
    
    test('card with header snapshot', () => {
      const header = <div>Header Content</div>;
      const { container } = render(
        <Card header={header}>
          <p>Card with header content</p>
        </Card>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
    
    test('card with footer snapshot', () => {
      const footer = <div>Footer Content</div>;
      const { container } = render(
        <Card footer={footer}>
          <p>Card with footer content</p>
        </Card>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
    
    test('loading card snapshot', () => {
      const { container } = render(
        <Card isLoading>
          <p>Loading card content</p>
        </Card>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });
}); 