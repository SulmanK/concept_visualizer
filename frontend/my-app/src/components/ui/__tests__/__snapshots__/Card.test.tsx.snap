// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card Component > Snapshots > card with footer snapshot 1`] = `
<div
  class="card"
>
  <div
    class="p-4 sm:p-6"
  >
    <p>
      Card with footer content
    </p>
  </div>
  <div
    class="px-4 py-3 sm:px-6 border-t border-indigo-200 bg-indigo-50/50"
  >
    <div>
      Footer Content
    </div>
  </div>
</div>
`;

exports[`Card Component > Snapshots > card with header snapshot 1`] = `
<div
  class="card"
>
  <div
    class="px-4 py-3 sm:px-6 border-b border-indigo-200 bg-indigo-50/50"
  >
    <div>
      Header Content
    </div>
  </div>
  <div
    class="p-4 sm:p-6"
  >
    <p>
      Card with header content
    </p>
  </div>
</div>
`;

exports[`Card Component > Snapshots > default card snapshot 1`] = `
<div
  class="card"
>
  <div
    class="p-4 sm:p-6"
  >
    <p>
      Default card content
    </p>
  </div>
</div>
`;

exports[`Card Component > Snapshots > elevated variant card snapshot 1`] = `
<div
  class="card bg-white border-none shadow-lg"
>
  <div
    class="p-4 sm:p-6"
  >
    <p>
      Elevated card content
    </p>
  </div>
</div>
`;

exports[`Card Component > Snapshots > gradient variant card snapshot 1`] = `
<div
  class="card-gradient"
>
  <div
    class="p-4 sm:p-6"
  >
    <p>
      Gradient card content
    </p>
  </div>
</div>
`;

exports[`Card Component > Snapshots > loading card snapshot 1`] = `
<div
  class="card"
>
  <div
    class="p-4 sm:p-6"
  >
    <div
      class="flex justify-center items-center py-12"
    >
      <div
        class="animate-spin rounded-full h-12 w-12 border-t-2 border-b-2 border-primary-600"
      />
    </div>
  </div>
</div>
`;
